x-pack/plugins/uptime/server/lib/adapters/monitors/elasticsearch_monitors_adapter.ts

```diff
@@ -332,44 +336,76 @@ export class ElasticsearchMonitorsAdapter implements UMMonitorsAdapter {
             },
           },
         },
+        size: 0,
         aggs: {
           id: {
             terms: {
               field: 'monitor.id',
+              size: 1000,
+            },
+            aggs: {
+              latest: {
+                top_hits: {
+                  size: 1,
```

Comment:
This would be up to `1000` fetches. But, perhaps more importantly, we're abusing top_hits fetch a single field per `monitor.id`.

I wonder if this request would be better served as

```js
{
  body: {
    _source: [ 'monitor.id', 'monitor.type', 'url.full', 'url.port' ],
    size: 1000,
    query: {
      range: {
        '@timestamp': {
          gte: dateRangeStart,
          lte: dateRangeEnd,
        },
       },
    },
    collapse: {
      field: 'monitor.id'
    },
    sort: {
      '@timestamp': 'desc'
    }
  }
}
```

Then we can simply iterate across the 1000 most recent, unique `monitor.id`'s values. In fairness, that is subtly different behavior from the aggregations. The `terms` aggregations would collect `1000` of the most common values per selected field -- the field collapsing would merely fetch the `1000` most recent (in both cases assuming >1000 exist) and build from there. But it also means fewer implied loops: the aggregations get looped over _per_ aggregation to convert to keys, whereas we'd just need one loop.

```js
const ids: any[] = [];
const ports: Set = new Set();
const types: Set = new Set();

const hits = get(result, 'hits.hits', []);
hits.forEach(hit => {
  const source: any = hit._source;
  ids.push({ id: source.monitor.id, url: url.full });
  ports.add(source.url.port);
  types.add(source.monitor.type);
});

return {
  id: ids,
  port: Array.from(ports),
  scheme: Array.from(types),
  status: [ 'up', 'down' ],
};
```

Reply:
I think this is much nicer/cleaner. Implemented here if you want to review: https://github.com/elastic/kibana/pull/29374/commits/f4f863d09644a5fabac24c45f5f1cb0fcb2a275f.